agent/cache/cache.go

```diff
@@ -819,14 +866,28 @@ func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ign
 				return
 			}
 
+			select {
+			case <-handle.stopCh:
+				return
+			default:
+			}
+
 			// If we're over the attempt minimum, start an exponential backoff.
 			if wait := backOffWait(attempt); wait > 0 {
-				time.Sleep(wait)
```

Comment:
After the initial refactor it turned out that the goroutine leak was **back!** 

Then I changed `time.Sleep` into cancellable ones via the `select` pattern and the leak went away.